packages/astro/src/core/build/plugins/plugin-analyzer.ts

```diff
@@ -17,6 +19,102 @@ function isPropagatedAsset(id: string) {
+		return 'dynamic';
+	}
+
+	const imports: Array<ImportDeclaration> = [];
+	const exports: Array<ExportNamedDeclaration | ExportDefaultDeclaration> = [];
+	walk(parentInfo.ast, {
+		enter(node) {
+			if (node.type === 'ImportDeclaration') {
+				imports.push(node as ImportDeclaration);
+			} else if (
+				node.type === 'ExportDefaultDeclaration' ||
+				node.type === 'ExportNamedDeclaration'
+			) {
+				exports.push(node as ExportNamedDeclaration | ExportDefaultDeclaration);
+			}
+		},
+	});
+	// All of the aliases the current component is imported as
+	const importNames: string[] = [];
+	// All of the aliases the child component is exported as
+	const exportNames: string[] = [];
+
+	for (const node of imports) {
+		const resolved = await this.resolve(node.source.value as string, parentInfo.id);
+		if (!resolved || resolved.id !== childInfo.id) continue;
+		for (const specifier of node.specifiers) {
+			// TODO: handle these?
+			if (specifier.type === 'ImportNamespaceSpecifier') continue;
+			const name =
+				specifier.type === 'ImportDefaultSpecifier' ? 'default' : specifier.imported.name;
+			// If we're importing the thing that the child exported, store the local name of what we imported
+			if (childExportNames.includes(name)) {
+				importNames.push(specifier.local.name);
+			}
+		}
+	}
+	for (const node of exports) {
+		if (node.type === 'ExportDefaultDeclaration') {
+			if (node.declaration.type === 'Identifier' && importNames.includes(node.declaration.name)) {
+				exportNames.push('default');
+				// return
+			}
+		} else {
+			// handle `export { x } from 'something';`, where the export and import are in the same node
+			if (node.source) {
+				const resolved = await this.resolve(node.source.value as string, parentInfo.id);
+				if (!resolved || resolved.id !== childInfo.id) continue;
+				for (const specifier of node.specifiers) {
+					if (childExportNames.includes(specifier.local.name)) {
+						importNames.push(specifier.local.name);
+						exportNames.push(specifier.exported.name);
+					}
+				}
+			}
+			if (node.declaration) {
+				if (node.declaration.type !== 'VariableDeclaration') continue;
+				for (const declarator of node.declaration.declarations) {
+					if (declarator.init?.type !== 'Identifier') continue;
+					if (declarator.id.type !== 'Identifier') continue;
+					if (importNames.includes(declarator.init.name)) {
+						exportNames.push(declarator.id.name);
+					}
+				}
+			}
+			for (const specifier of node.specifiers) {
+				if (importNames.includes(specifier.local.name)) {
+					exportNames.push(specifier.exported.name);
+				}
+			}
+		}
+	}
+	if (!importNames.length) return 'no';
+
+	// If the component is imported by another component, assume it's in use
+	// and start tracking this new component now
+	if (parentInfo.id.endsWith('.astro')) {
+		exportNames.push('default');
+	} else if (parentInfo.id.endsWith('.mdx')) {
```

Comment:
I assume this is unacceptable - mdx logic shouldn't be in the core. Unsure on how to resolve.

Reply:
Yeah, given we already have this in the code. Not ideal.

Reply:
I haven't tested this yet, but shouldn't the `ast` already contain these exports and added them to `exportNames`?

Reply:
This is to handle a slightly different case. If component `Title.astro` is imported and used by `TitleBar.astro`, `TitleBar.astro` is not going to re-export `Title`. Therefore, this line basically switches from tracking Title to TitleBar, so the page will include `Title`'s scripts.